File: wire-ios-mocktransport/Source/Conversations/MockTransportSession+conversations.m
//
// Wire
// Copyright (C) 2024 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

@import WireTransport;
@import WireUtilities;
@import CoreData;
@import WireProtos;

#import "MockTransportSession+conversations.h"
#import <WireMockTransport/WireMockTransport-Swift.h>
#import "MockTransportSession+assets.h"
#import "MockTransportSession+OTR.h"
#import "NSManagedObjectContext+executeFetchRequestOrAssert.h"


static char* const ZMLogTag ZM_UNUSED = "MockTransport";

@implementation MockTransportSession (Conversations)

// swiftlint:disable:next todo_requires_jira_link
//TODO: filter requests using array of NSPredicates

// handles /conversations
- (ZMTransportResponse *)processConversationsRequest:(ZMTransportRequest *)request;
{
    if ([request matchesWithPath:@"/conversations" method:ZMTransportRequestMethodGet]) {
        return [self processConversationsGetConversationsIDs:request.queryParameters[@"ids"] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/ids" method:ZMTransportRequestMethodGet])
    {
        return [self processConversationIDsQuery:request.queryParameters apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/join" method:ZMTransportRequestMethodGet])
    {
        return [self processFetchConversationIdAndNameWith:request.queryParameters apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*" method:ZMTransportRequestMethodGet])
    {
        return [self processConversationsGetConversation:[request RESTComponentAtIndex:1] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations" method:ZMTransportRequestMethodPost])
    {
        return [self processConversationsPostConversationsRequest:request];
    }
    else if ([request matchesWithPath:@"/conversations/*/otr/messages" method:ZMTransportRequestMethodPost])
    {
        if (request.binaryData != nil) {
            return [self processAddOTRMessageToConversation:[request RESTComponentAtIndex:1]
                                          withProtobuffData:request.binaryData
                                                      query:request.queryParameters
                                                 apiVersion:request.apiVersion];
        }
        else {
            return [self processAddOTRMessageToConversation:[request RESTComponentAtIndex:1]
                                                    payload:[request.payload asDictionary]
                                                      query:request.queryParameters
                                                 apiVersion:request.apiVersion];
        }
    }
    else if ([request matchesWithPath:@"/conversations/*/members" method:ZMTransportRequestMethodPost])
    {
        return [self processAddMembersToConversation:[request RESTComponentAtIndex:1] payload:[request.payload asDictionary] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*" method:ZMTransportRequestMethodPut])
    {
        return [self processPutConversation:[request RESTComponentAtIndex:1] payload:[request.payload asDictionary] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*/members/*" method:ZMTransportRequestMethodPut])
    {
        return [self processPutMembersInConversation:[request RESTComponentAtIndex:1] member:[request RESTComponentAtIndex:3] payload:[request.payload asDictionary] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*/self" method:ZMTransportRequestMethodPut])
    {
        return [self processPutConversationSelf:[request RESTComponentAtIndex:1] payload:[request.payload asDictionary] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*/members/*" method:ZMTransportRequestMethodDelete])
    {
        return [self processDeleteConversation:[request RESTComponentAtIndex:1] member:[request RESTComponentAtIndex:3] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*/typing" method:ZMTransportRequestMethodPost])
    {
        return [self processConversationTyping:[request RESTComponentAtIndex:1] payload:[request.payload asDictionary] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*/assets/*" method:ZMTransportRequestMethodGet])
    {
        return [self processAssetRequest:request];
    }
    else if ([request matchesWithPath:@"/conversations/*/otr/assets/*" method:ZMTransportRequestMethodGet])
    {
        return [self processAssetRequest:request];
    }
    else if ([request matchesWithPath:@"/conversations/*/bots" method:ZMTransportRequestMethodPost]) {
        return [self processServiceRequest:request];
    }
    else if ([request matchesWithPath:@"/conversations/*/bots/*" method:ZMTransportRequestMethodDelete]) {
        return [self processDeleteBotRequest:request];
    }
    else if ([request matchesWithPath:@"/conversations/*/access" method:ZMTransportRequestMethodPut]) {
        return [self processAccessModeUpdateForConversation:[request RESTComponentAtIndex:1] payload:[request.payload asDictionary] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*/code" method:ZMTransportRequestMethodGet]) {
        return [self processFetchLinkForConversation:[request RESTComponentAtIndex:1] payload:[request.payload asDictionary] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*/code" method:ZMTransportRequestMethodPost]) {
        return [self processCreateLinkForConversation:[request RESTComponentAtIndex:1] payload:[request.payload asDictionary] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*/code" method:ZMTransportRequestMethodDelete]) {
        return [self processDeleteLinkForConversation:[request RESTComponentAtIndex:1] payload:[request.payload asDictionary] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*/features/conversationGuestLinks" method:ZMTransportRequestMethodGet]) {
        return [self processGuestLinkFeatureStatusForConversation:[request RESTComponentAtIndex:1] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*/receipt-mode" method:ZMTransportRequestMethodPut]) {
        return [self processReceiptModeUpdateForConversation:[request RESTComponentAtIndex:1] payload:[request.payload asDictionary] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/*/roles" method:ZMTransportRequestMethodGet]) {
        return [self processFetchRolesForConversation:[request RESTComponentAtIndex:1] payload:[request.payload asDictionary] apiVersion:request.apiVersion];
    }
    else if ([request matchesWithPath:@"/conversations/join" method:ZMTransportRequestMethodPost]) {
        return [self processJoinConversationWithPayload:[request.payload asDictionary] apiVersion:request.apiVersion];
    }

    return [ZMTransportResponse responseWithPayload:nil HTTPStatus:404 transportSessionError:nil apiVersion:request.apiVersion];

}


// POST /conversations/<id>/otr/messages
- (ZMTransportResponse *)processAddOTRMessageToConversation:(NSString *)conversationId payload:(NSDictionary *)payload query:(NSDictionary *)query apiVersion:(APIVersion)apiVersion;
{
    NSAssert(self.selfUser != nil, @"No self user in mock transport session");
    
    MockConversation *conversation = [self fetchConversationWithIdentifier:conversationId];
    NSAssert(conversation, @"No conv found");

    NSDictionary *recipients = payload[@"recipients"];
    MockUserClient *senderClient = [self otrMessageSender:payload];
    if (senderClient == nil) {
        return [ZMTransportResponse responseWithPayload:nil HTTPStatus:404 transportSessionError:nil apiVersion:apiVersion];
    }

    NSString *onlyForUser = query[@"report_missing"];
    NSDictionary *missedClients = [self missedClients:recipients conversation:conversation sender:senderClient onlyForUserId:onlyForUser];
    NSDictionary *deletedClients = [self deletedClients:recipients conversation:conversation];
    
    NSDictionary *responsePayload = @{@"redundant": @{}, @"missing": missedClients, @"deleted": deletedClients, @"time": [NSDate date].transportString};

    NSInteger statusCode = 412;
    if (missedClients.count == 0) {
        statusCode = 201;
        [self insertOTRMessageEventsToConversation:conversation requestPayload:payload createEventBlock:^MockEvent *(MockUserClient *recipient, NSData *messageData) {
            return [conversation insertOTRMessageFromClient:senderClient toClient:recipient data:messageData];
        }];
    }
    
    return [ZMTransportResponse responseWithPayload:responsePayload HTTPStatus:statusCode transportSessionError:nil apiVersion:apiVersion];
}

- (ZMTransportResponse *)processPutConversation:(NSString *)conversationId payload:(NSDictionary *)payload apiVersion:(APIVersion)apiVersion;
{
    MockConversation *conversation = [self conversationByIdentifier:conversationId];
    if (conversation == nil) {
        return [ZMTransportResponse responseWithPayload:nil HTTPStatus:404 transportSessionError:nil apiVersion:apiVersion];
    }
    
    NSString *newName = [payload optionalStringForKey:@"name"];
    if (newName == nil) {
        return [ZMTransportResponse responseWithPayload:@{@"error":@"no name in payload"} HTTPStatus:400 transportSessionError:nil apiVersion:apiVersion];
    }
    
    NSNumber *receiptMode = [payload optionalNumberForKey:@"receipt_mode"];
    
    if (receiptMode != nil) {
        [conversation changeReceiptModeByUser:self.selfUser receiptMode:receiptMode.intValue];
    }
    
    MockEvent *event = [conversation changeNameByUser:self.selfUser name:newName];
    return [ZMTransportResponse responseWithPayload:event.transportData HTTPStatus:200 transportSessionError:nil apiVersion:apiVersion];
}

- (ZMTransportResponse *)processPutMembersInConversation:(NSString *)conversationId member:(NSString *)memberId payload:(NSDictionary *)payload apiVersion:(APIVersion)apiVersion;
{
    MockConversation *conversation = [self conversationByIdentifier:conversationId];
    if (conversation == nil) {
        return [ZMTransportResponse responseWithPayload:nil HTTPStatus:404 transportSessionError:nil apiVersion:apiVersion];
    }
    
    NSString *conversationRole = [payload optionalStringForKey:@"conversation_role"];
    if (conversationRole == nil) {
        return [ZMTransportResponse responseWithPayload:@{@"error":@"no conversation_role in payload"} HTTPStatus:400 transportSessionError:nil apiVersion:apiVersion];
    }
    
    MockUser *user = [self fetchUserWithIdentifier:memberId];
    if(user == nil) {
        return [ZMTransportResponse responseWithPayload:@{@"code" : @400, @"message": @"Unknown user", @"label": @""} HTTPStatus:400 transportSessionError:nil apiVersion:apiVersion];
    }
    
    if (![conversation.activeUsers containsObject:user]) {
        return [ZMTransportResponse responseWithPayload:nil HTTPStatus:403 transportSessionError:nil apiVersion:apiVersion];
    }
    MockParticipantRole * participantRoleMember = [MockParticipantRole insertIn:self.managedObjectContext conversation:conversation user:user];
    participantRoleMember.role = [conversationRole isEqualToString:MockConversation.member] ? MockRole.memberRole : MockRole.adminRole;
    
    MockParticipantRole * participantRoleSelfUser = [MockParticipantRole insertIn:self.managedObjectContext conversation:conversation user:self.selfUser];
    participantRoleSelfUser.role = MockRole.adminRole;
    
    return [ZMTransportResponse responseWithPayload:nil HTTPStatus:200 transportSessionError:nil apiVersion:apiVersion];
}


// returns YES if the payload contains "muted" information
- (BOOL)updateConversation:(MockConversation *)conversation isOTRMutedFromPutSelfConversationPayload:(NSDictionary *)payload
{
    NSString *mutedRef = [payload optionalStringForKey:@"otr_muted_ref"];
    if (mutedRef != nil) {
        NSNumber *muted = [payload optionalNumberForKey:@"otr_muted"];
        conversation.otrMuted = ([muted isEqual:@1]);
        conversation.otrMutedRef = mutedRef;
    }
    
    return mutedRef != nil;
}

// returns YES if the payload contains "muted" information
- (BOOL)updateConversation:(MockConversation *)conversation isOTRArchivedFromPutSelfConversationPayload:(NSDictionary *)payload
{
    NSString *archivedRef = [payload optionalStringForKey:@"otr_archived_ref"];
    if (archivedRef != nil) {
        NSNumber *archived = [payload optionalNumberForKey:@"otr_archived"];
        conversation.otrArchived = ([archived isEqual:@1]);
        conversation.otrArchivedRef = archivedRef;
    }
    
    return archivedRef != nil;
}

- (BOOL)updateConversation:(MockConversation *)conversation isOTRMutedStatusFromPutSelfConversationPayload:(NSDictionary *)payload
{
    NSNumber *mutedStatus = [payload optionalNumberForKey:@"otr_muted_status"];
    conversation.otrMutedStatus = mutedStatus;
    return mutedStatus != nil;
}

- (ZMTransportResponse *)processPutConversationSelf:(NSString *)conversationId payload:(NSDictionary *)payload apiVersion:(APIVersion)apiVersion;
{
    MockConversation *conversation = [self conversationByIdentifier:conversationId];
    if (conversation == nil) {
        return [ZMTransportResponse responseWithPayload:nil HTTPStatus:404 transportSessionError:nil apiVersion:apiVersion];
    }
    
    BOOL hadOTRMuted = [self updateConversation:conversation isOTRMutedFromPutSelfConversationPayload:payload];
    BOOL hadOTRArchived = [self updateConversation:conversation isOTRArchivedFromPutSelfConversationPayload:payload];
    BOOL hadOTRMutedStatus = [self updateConversation:conversation isOTRMutedStatusFromPutSelfConversationPayload:payload];

    if( !hadOTRArchived && !hadOTRMuted && !hadOTRMutedStatus) {
        return [ZMTransportResponse responseWithPayload:@{@"error":@"no useful payload"} HTTPStatus:400 transportSessionError:nil apiVersion:apiVersion];
    }
    
    return [ZMTransportResponse responseWithPayload:nil HTTPStatus:200 transportSessionError:nil apiVersion:apiVersion];
}

- (ZMTransportResponse *)processConversationsGetConversationsIDs:(NSString *)ids apiVersion:(APIVersion)apiVersion;
{
    NSFetchRequest *request = [MockConversation sortedFetchRequest];
    NSArray *conversations = [self.managedObjectContext executeFetchRequestOrAssert_mt:request];
    NSMutableArray *data = [NSMutableArray array];
    
    if (ids != nil) {
        
        NSSet *requestedIDs = [NSSet setWithArray:[ids componentsSeparatedByString:@","]];
        
        for (MockConversation *conversation in conversations) {
            if([requestedIDs containsObject:conversation.identifier]) {
                [data addObject:conversation.transportData];
            }
        }
    }
    else {
        for (MockConversation *conversation in conversations) {
            [data addObject:conversation.transportData];
        }
    }
    
    return [ZMTransportResponse responseWithPayload:@{@"conversations":data} HTTPStatus:200 transportSessionError:nil apiVersion:apiVersion];

}

- (ZMTransportResponse *)processConversationsGetConversation:(NSString *)conversationId apiVersion:(APIVersion)apiVersion;
{
    MockConversation *conversation = [self conversationByIdentifier:conversationId];
    
    if (conversation == nil) {
        return [ZMTransportResponse responseWithPayload:nil HTTPStatus:404 transportSessionError:nil apiVersion:apiVersion];
    } else if (![conversation.activeUsers containsObject:self.selfUser]) {
        return [ZMTransportResponse responseWithPayload:nil HTTPStatus:403 transportSessionError:nil apiVersion:apiVersion];
    }
    
    return [ZMTransportResponse responseWithPayload:conversation.transportData HTTPStatus:200 transportSessionError:nil apiVersion:apiVersion];
}

- (ZMTransportResponse *)processConversationsPostConversationsRequest:(ZMTransportRequest *)request;
{
    NSArray *participantIDs = request.payload.asDictionary[@"users"];
    NSString *name = request.payload.asDictionary[@"name"];
    NSString *conversationRole = request.payload.asDictionary[@"conversation_role"];

    
    NSMutableArray *otherUsers = [NSMutableArray array];
    
    for (NSString *id in participantIDs) {
        NSPredicate *predicate = [NSPredicate predicateWithFormat:@"identifier == %@", id];
        NSFetchRequest *fetchRequest = [MockUser sortedFetchRequestWithPredicate:predicate];
        
        NSArray *results = [self.managedObjectContext executeFetchRequestOrAssert_mt:fetchRequest];

        if (results.count == 1) {
            MockUser *user = results[0];
            [otherUsers addObject:user];
        }
    }
    
    MockConversation *conversation = [self insertGroupConversationWithSelfUser:self.selfUser otherUsers:otherUsers];
    if(name != nil) {
        [conversation changeNameByUser:self.selfUser name:name];
    }    
    for (MockUser *user in otherUsers) {
        MockParticipantRole * participantRoleMember = [MockParticipantRole insertIn:self.managedObjectContext conversation:conversation user:user];
        participantRoleMember.role = [conversationRole isEqualToString:MockConversation.member] ? MockRole.memberRole : MockRole.adminRole;
    }
    MockParticipantRole * participantRoleSelfUser = [MockParticipantRole insertIn:self.managedObjectContext conversation:conversation user:self.selfUser];
    participantRoleSelfUser.role = MockRole.adminRole;
    
    return [ZMTransportResponse responseWithPayload:[conversation transportData] HTTPStatus:200 transportSessionError:nil apiVersion:request.apiVersion];
}


- (ZMTransportResponse *)processDeleteConversation:(NSString *)conversationId member:(NSString *)memberId apiVersion:(APIVersion)apiVersion;
{
    MockConversation *conversation = [self fetchConversationWithIdentifier:conversationId];
    if (conversation == nil) {
        return [ZMTransportResponse responseWithPayload:nil HTTPStatus:404 transportSessionError:nil apiVersion:apiVersion];
    }
    
    MockUser *user = [self fetchUserWithIdentifier:memberId];
    MockEvent *event = [conversation removeUsersByUser:self.selfUser removedUser:user];
    
    return [ZMTransportResponse responseWithPayload:event.transportData HTTPStatus:200 transportSessionError:nil apiVersion:apiVersion];
}


- (ZMTransportResponse *)processAddMembersToConversation:(NSString *)conversationId payload:(NSDictionary *)payload apiVersion:(APIVersion)apiVersion;
{
    MockConversation *conversation = [self fetchConversationWithIdentifier:conversationId];
    
    NSArray *addedUserIDs = payload[@"users"];
    NSString *conversationRole = payload[@"conversation_role"];
    NSMutableArray *addedUsers = [NSMutableArray array];
    MockUser *selfUser = self.selfUser;
    NSAssert(selfUser != nil, @"Self not found");
    for (NSString *userID in addedUserIDs) {
        MockUser *user = [self fetchUserWithIdentifier:userID];
        if(user == nil) {
            return [ZMTransportResponse responseWithPayload:@{@"code" : @403, @"message": @"Unknown user", @"label": @""} HTTPStatus:403 transportSessionError:nil apiVersion:apiVersion];
        }
        MockParticipantRole * participantRoleMember = [MockParticipantRole insertIn:self.managedObjectContext conversation:conversation user:user];
        participantRoleMember.role = [conversationRole isEqualToString:MockConversation.member] ? MockRole.memberRole : MockRole.adminRole;
        
        MockConnection *connection = [self fetchConnectionFrom:selfUser to:user];
        if (connection == nil) {
            return [ZMTransportResponse responseWithPayload:@{@"code" : @403, @"message": @"Requestor is not connected to users invited", @"label": @"" } HTTPStatus:403 transportSessionError:nil apiVersion:apiVersion];
        }
        [addedUsers addObject:user];
    }
    
    MockEvent *event = [conversation addUsersByUser:self.selfUser addedUsers:addedUsers];
    return [ZMTransportResponse responseWithPayload:event.transportData HTTPStatus:200 transportSessionError:nil apiVersion:apiVersion];
}

- (MockConversation *)conversationByIdentifier:(NSString *)identifier
{
    NSFetchRequest *request = [MockConversation sortedFetchRequestWithPredicate:[NSPredicate predicateWithFormat:@"identifier == %@", identifier]];
    
    NSArray *conversations = [self.managedObjectContext executeFetchRequestOrAssert_mt:request];
    RequireString(conversations.count <= 1, "Too many conversations with one identifier");
    
    return conversations.count > 0 ? conversations[0] : nil;
}

// POST /conversations/<id>/typing
- (ZMTransportResponse *)processConversationTyping:(NSString *)conversationId payload:(NSDictionary *)payload apiVersion:(APIVersion)apiVersion
{
    MockConversation *conversation = [self conversationByIdentifier:conversationId];
    if (conversation == nil) {
        return [ZMTransportResponse responseWithPayload:nil HTTPStatus:404 transportSessionError:nil apiVersion:apiVersion];
    }
    
    BOOL isTyping = [[payload optionalStringForKey:@"status"] isEqualToString:@"started"];
    MockEvent *event = [conversation insertTypingEventFromUser:self.selfUser isTyping:isTyping];
    NSDictionary *responsePayload = [event transportData].asDictionary;
    
    return [ZMTransportResponse responseWithPayload:responsePayload HTTPStatus:201 transportSessionError:nil apiVersion:apiVersion];
}

// GET /conversations/ids
- (ZMTransportResponse *)processConversationIDsQuery:(NSDictionary *)query apiVersion:(APIVersion)apiVersion
{
    NSString *sizeString = [query optionalStringForKey:@"size"];
    NSUUID *start = [query optionalUuidForKey:@"start"];
    
    NSFetchRequest *request = [MockConversation sortedFetchRequest];
    NSArray *conversations = [self.managedObjectContext executeFetchRequestOrAssert_mt:request];

    NSArray *conversationIDs = [conversations mapWithBlock:^id(MockConversation *obj) {
        return obj.identifier;
    }];
    
    if(start != nil) {
        NSUInteger index = [conversationIDs indexOfObject:start.transportString];
        if(index != NSNotFound) {
            conversationIDs = [conversationIDs subarrayWithRange:NSMakeRange(index+1, conversationIDs.count - index-1)];
        }
    }

    BOOL hasMore = NO;
    if(sizeString != nil) {
        NSUInteger remainingConversations = conversationIDs.count;
        NSUInteger pageSize = (NSUInteger) sizeString.integerValue;
        hasMore = (remainingConversations > pageSize);
        NSUInteger numOfConversations = MIN(remainingConversations, pageSize);
        conversationIDs = [conversationIDs subarrayWithRange:NSMakeRange(0u, numOfConversations)];
    }
    
    NSDictionary *payload = @{
                              @"has_more": @(hasMore),
                              @"conversations": conversationIDs
                              };
    
    ZMTransportResponse *response = [ZMTransportResponse responseWithPayload:payload HTTPStatus:200 transportSessionError:nil apiVersion:apiVersion];
    return response;
}

@end

File: Source/Conversations/MockTransportSession+conversations.swift
////
// Wire
// Copyright (C) 2018 Wire Swiss GmbH
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program. If not, see http://www.gnu.org/licenses/.
//

import Foundation
import WireProtos

extension MockTransportSession {

    @objc(fetchConversationWithIdentifier:)
    public func fetchConversation(with identifier: String) -> MockConversation? {
        let request = MockConversation.sortedFetchRequest()
        request.predicate = NSPredicate(format: "identifier == %@", identifier.lowercased())
        let conversations = managedObjectContext.executeFetchRequestOrAssert(request) as? [MockConversation]
        return conversations?.first
    }
    
    @objc(processReceiptModeUpdateForConversation:payload:)
    public func processReceiptModeUpdate(for conversationId: String, payload: [String:  AnyHashable]) -> ZMTransportResponse {
        guard let conversation = fetchConversation(with: conversationId) else {
            return ZMTransportResponse(payload: nil, httpStatus: 404, transportSessionError: nil)
        }
        guard let receiptMode = payload["receipt_mode"] as? Int else {
            return ZMTransportResponse(payload: nil, httpStatus: 400, transportSessionError: nil)
        }
        guard receiptMode != conversation.receiptMode?.intValue else {
            return ZMTransportResponse(payload: nil, httpStatus: 204, transportSessionError: nil)
        }
        
        conversation.receiptMode = NSNumber(value: receiptMode)
        
        let responsePayload = [
            "conversation" : conversation.identifier,
            "type" : "conversation.receipt-mode-update",
            "time" : NSDate().transportString(),
            "from" : selfUser.identifier,
            "data" : ["receipt_mode": receiptMode]] as ZMTransportData
        
        return ZMTransportResponse(payload: responsePayload, httpStatus: 200, transportSessionError: nil)
    }

    @objc(processAccessModeUpdateForConversation:payload:)
    public func processAccessModeUpdate(for conversationId: String, payload: [String : AnyHashable]) -> ZMTransportResponse {
        guard let conversation = fetchConversation(with: conversationId) else {
            return ZMTransportResponse(payload: nil, httpStatus: 404, transportSessionError: nil)
        }
        guard let accessRole = payload["access_role"] as? String else {
            return ZMTransportResponse(payload: nil, httpStatus: 400, transportSessionError: nil)
        }
        guard let access = payload["access"] as? [String] else {
            return ZMTransportResponse(payload: nil, httpStatus: 400, transportSessionError: nil)
        }

        conversation.accessRole = accessRole
        conversation.accessMode = access

        let responsePayload = [
            "conversation" : conversation.identifier,
            "type" : "conversation.access-update",
            "time" : NSDate().transportString(),
            "from" : selfUser.identifier,
            "data" : [
                "access_role" : conversation.accessRole,
                "access" : conversation.accessMode
            ]
        ] as ZMTransportData
        return ZMTransportResponse(payload: responsePayload, httpStatus: 200, transportSessionError: nil)
    }
    
    @objc(processFetchLinkForConversation:payload:)
    public func processFetchLink(for conversationId: String, payload: [String: AnyHashable]) -> ZMTransportResponse {
        guard let conversation = fetchConversation(with: conversationId) else {
            return ZMTransportResponse(payload: nil, httpStatus: 404, transportSessionError: nil)
        }
        
        guard Set(conversation.accessMode) == Set(["invite", "code"]) else {
            return ZMTransportResponse(payload: ["label": "invalid-op"] as ZMTransportData, httpStatus: 403, transportSessionError: nil)
        }
        
        guard let link = conversation.link else {
            return ZMTransportResponse(payload: ["label": "no-conversation-code"] as ZMTransportData, httpStatus: 404, transportSessionError: nil)
        }
        
        return ZMTransportResponse(payload: ["uri": link,
                                             "key": "test-key",
                                             "code": "test-code"] as ZMTransportData, httpStatus: 200, transportSessionError: nil)
    }
    
    @objc(processFetchRolesForConversation:payload:)
    public func processFetchRoles(for conversationId: String, payload: [String: AnyHashable]) -> ZMTransportResponse {
        guard let conversation = fetchConversation(with: conversationId) else {
            return ZMTransportResponse(payload: nil, httpStatus: 404, transportSessionError: nil)
        }
        
        let roles = conversation.team?.roles ?? conversation.nonTeamRoles!
        let payload: [String : Any] = [
            "conversation_roles" : roles.map { $0.payload }
        ]
        return ZMTransportResponse(payload: payload as ZMTransportData, httpStatus: 200, transportSessionError: nil)
    }
    
    @objc(processCreateLinkForConversation:payload:)
    public func processCreateLink(for conversationId: String, payload: [String: AnyHashable]) -> ZMTransportResponse {
        guard let conversation = fetchConversation(with: conversationId) else {
            return ZMTransportResponse(payload: nil, httpStatus: 404, transportSessionError: nil)
        }
        
        guard Set(conversation.accessMode) == Set(["invite", "code"]) else {
            return ZMTransportResponse(payload: ["label": "invalid-op"] as ZMTransportData, httpStatus: 403, transportSessionError: nil)
        }
        
        // link already exists
        if let link = conversation.link {
            return ZMTransportResponse(payload: ["uri": link,
                                                 "key": "test-key",
                                                 "code": "test-code"] as ZMTransportData, httpStatus: 200, transportSessionError: nil)
        }
        // new link must be created
        else {
            let link = "https://wire-website.com/test-link"
            
            conversation.link = link
            
            let payload = [
                    "conversation" : conversationId,
                    "data" : [
                        "uri" : link,
                        "key" : "test-key",
                        "code" : "test-code"
                    ],
                    "type" : "conversation.code-update",
                    "time" : NSDate().transportString(),
                    "from" : selfUser.identifier
            ] as ZMTransportData
            return ZMTransportResponse(payload: payload, httpStatus: 201, transportSessionError: nil)
        }
        
    }
    
    @objc(processDeleteLinkForConversation:payload:)
    public func processDeleteLink(for conversationId: String, payload: [String: AnyHashable]) -> ZMTransportResponse {
        guard let conversation = fetchConversation(with: conversationId) else {
            return ZMTransportResponse(payload: nil, httpStatus: 404, transportSessionError: nil)
        }
        
        guard Set(conversation.accessMode) == Set(["invite", "code"]) else {
            return ZMTransportResponse(payload: ["label": "invalid-op"] as ZMTransportData, httpStatus: 403, transportSessionError: nil)
        }
        
        // link already exists
        if let _ = conversation.link {
            conversation.link = nil
            return ZMTransportResponse(payload: nil, httpStatus: 200, transportSessionError: nil)
        }
        // new link must be created
        else {
            return ZMTransportResponse(payload: nil, httpStatus: 403, transportSessionError: nil)
        }
    }

    /// Returns a response for the POST "/conversations/join" request
    /// - Parameter query: payload
    /// - Returns: response payload depending on code value:
    /// - "existing-conversation-code" -  response payload should have a httpStatus 204
    /// - "test-code" -  response payload should contain a new conversation
    /// - "wrong-code" - there should be an error in the response payload
    @objc(processJoinConversationWithPayload:)
    public func processJoinConversation(with payload: [String: AnyHashable]) -> ZMTransportResponse {
        guard let code = payload["code"] as? String else {
            let payload = ["label" : "no-conversation-code"] as ZMTransportData
            return ZMTransportResponse(payload: payload, httpStatus: 404, transportSessionError: nil)
        }

        switch code {
        case "existing-conversation-code":
            return ZMTransportResponse(payload: nil, httpStatus: 204, transportSessionError: nil)

        case "test-code":
            let creator = insertUserWithName(name: "Bob")
            let conversation = MockConversation.insert(into: managedObjectContext, creator: creator, otherUsers: [], type: .group)

            let responsePayload = [
                "conversation" : conversation.identifier,
                "type" : "conversation.member-join",
                "time" : NSDate().transportString(),
                "data": [
                    "users" : [
                        [
                            "conversation_role": "wire_member",
                            "id": selfUser.identifier
                        ]
                    ],
                    "user_ids": [
                        selfUser.identifier
                    ]
                ],
                "from" : selfUser.identifier] as ZMTransportData
            return ZMTransportResponse(payload: responsePayload, httpStatus: 200, transportSessionError: nil)

        default:
            let payload = ["label" : "no-conversation-code"] as ZMTransportData
            return ZMTransportResponse(payload: payload, httpStatus: 404, transportSessionError: nil)
        }
    }


    /// Returns a response for the GET "/conversations/join" request
    /// - Parameter query: request query
    /// - Returns: response payload containing conversation ID and name depending on code value:
    /// - "existing-conversation-code" -  response payload should contain an existing conversation with a selfUser
    /// - "test-code" -  response payload should contain a new conversation ID
    /// - "wrong-code" - there should be an error in the response payload
    @objc(processFetchConversationIdAndNameWith:)
    public func processFetchConversationIdAndName(with query: [String: AnyHashable]) -> ZMTransportResponse {
        guard let code = query["code"] as? String else {
            let payload = ["label" : "no-conversation-code"] as ZMTransportData
            return ZMTransportResponse(payload: payload, httpStatus: 404, transportSessionError: nil)
        }

        switch code {
        case "existing-conversation-code":
            let conversation = fetchConversation(selfUserIdentifier: selfUser.identifier)
            let responsePayload = [
                "id" : conversation!.identifier,
                "name" : "Test"] as ZMTransportData
            return ZMTransportResponse(payload: responsePayload, httpStatus: 200, transportSessionError: nil)

        case "test-code":
            let responsePayload = [
                "id" : UUID.create().transportString(),
                "name" : "Test"] as ZMTransportData
            return ZMTransportResponse(payload: responsePayload, httpStatus: 200, transportSessionError: nil)

        default:
            let payload = ["label" : "no-conversation-code"] as ZMTransportData
            return ZMTransportResponse(payload: payload, httpStatus: 404, transportSessionError: nil)
        }
    }

    @objc
    public func processAddOTRMessage(toConversation conversationID: String,
                                     withProtobuffData data: Data,
                                     query: [String: Any]) -> ZMTransportResponse {
        guard
            let conversation = fetchConversation(with: conversationID),
            let otrMetaData = try? NewOtrMessage(serializedData: data),
            let senderClient = otrMessageSender(fromClientId: otrMetaData.sender) else {
                return ZMTransportResponse(payload: nil, httpStatus: 404, transportSessionError: nil)
        }
        
        var onlyForUser = query["report_missing"] as? String
        if otrMetaData.reportMissing.count > 0, let userId = otrMetaData.reportMissing.first {
            onlyForUser = UUID(data: userId.uuid)?.transportString()
        }
        
        let missedClients = self.missedClients(fromRecipients: otrMetaData.recipients, conversation: conversation, sender: senderClient, onlyForUserId: onlyForUser)
        let deletedClients = self.deletedClients(fromRecipients: otrMetaData.recipients, conversation: conversation)
        
        let payload: [String: Any] = [
            "missing": missedClients,
            "deleted": deletedClients,
            "time": Date().transportString()
        ]
        
        var statusCode = 412
        if missedClients.isEmpty {
            statusCode = 201
            insertOTRMessageEvents(
                toConversation: conversation,
                recipients: otrMetaData.recipients,
                senderClient: senderClient,
                createEventBlock: { recipient, messageData, decryptedData in
                    let event = conversation.insertOTRMessage(from: senderClient, to: recipient, data: messageData)
                    event.decryptedOTRData = decryptedData
                    return event
                })
        }
        
        return ZMTransportResponse(payload: payload as ZMTransportData, httpStatus: statusCode, transportSessionError: nil)
    }

    private func fetchConversation(selfUserIdentifier: String) -> MockConversation? {
        let request = MockConversation.sortedFetchRequest()
        request.predicate = NSPredicate(format: "selfIdentifier == %@", selfUserIdentifier.lowercased())
        let conversations = managedObjectContext.executeFetchRequestOrAssert(request) as? [MockConversation]
        return conversations?.first
    }
}
